Increased compile time max GPUs to 512. Switched to int16_t DeviceIndex.

tringwald opened the pull request:
Fixes #115331.

This PR increases the number of valid GPU devices to 512 (from 64) in order to future-proof PyTorch for providers that offer [single nodes with a large device count](https://www.tensorwave.com/). Until now, `DeviceIndex` was an `int8_t`, thus multiple changes were necessary:

- `DeviceIndex` changed to `int16_t`. Updated consumers that assume it to be an `int8_t`.
- Updated bounds checking for `torch.device()` in the Python frontend. Right now, we allow funny things like `torch.device('cpu', 200).index == -56`, which is undefined behavior. I inserted some checks to only allow values between 0 and `c10::Device::MAX_NUM_DEVICES - 1`.
- Updated the `ArgumentInfo` struct as it hardcodes the device index as 8 bit field [^1]. Might be a breaking change, not sure if users rely on this.
- Introduced `c10::Device::MAX_NUM_DEVICES` as a replacement for the old `C10_COMPILE_TIME_MAX_GPUS`


cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @penguinwu @tianyu-l @yf225

[^1]: This field was unsigned, so I guess this has also been undef behavior the whole time? Our default device index is -1, so this always wrapped around to 255 when written to the `ArgumentInfo` struct. When I switched the `DeviceIndex` to `int16_t`, it actually stayed 255 after unpacking from `ArgumentInfo` again, as the `DeviceIndex` was now wide enough that it didn't wrap back to -1.

tringwald committed:
Increased compile time max GPUs to 512. Switched to int16_t DeviceIndex. Changed some core JIT structures to accommodate the new 16 bit DeviceIndex. Added tests. Updated bounds checks.

tringwald commented:
Current CI run failures seem to come from `main`.

albanD commented on c10/core/Device.h:
```diff
@@ -16,7 +17,10 @@ namespace c10 {
+using DeviceIndex = int16_t;
+
+/// The last, still valid device index
+constexpr int64_t MAX_DEVICE_INDEX = std::numeric_limits<DeviceIndex>::max();
```
Why not use 512 here and set C10_COMPILE_TIME_MAX_GPUS to this value?

tringwald replied:
I was thinking about this too, but decided not to couple the amount of GPUs to the max possible index in case we want to support things like `meta` and `lazy` tensors with >512 indices. I don't know if that's a realistic use-case though, I can also just reuse `C10_COMPILE_TIME_MAX_GPUS`.

albanD replied:
I think the same argument for limiting max GPUs for now applies here. I'm sure someone will soon use this constant to store a static array :p 
It's an easy one to flip so I would prefer all the devices to be consistent.

tringwald replied:
I now introduced a new constant `C10_MAX_NUM_DEVICES` and set `C10_COMPILE_TIME_MAX_GPUS` based on it. Maybe we should just remove `C10_COMPILE_TIME_MAX_GPUS` completely from `CUDAMacros.h` and replace everything with `C10_MAX_NUM_DEVICES`.

albanD commented on c10/core/TensorImpl.h:
```diff
@@ -3169,7 +3169,7 @@ class C10_TensorImpl_Size_Check_Dummy_Class : private TensorImpl {
 #if UINTPTR_MAX == 0xFFFFFFFF
   // This is a 32-bit system
   static constexpr bool check_sizes() {
-    constexpr size_t tsize = 20 * sizeof(int64_t);
```
That sounds like too much? Is bad padding happening?

tringwald replied:
It's probably related  to the `optional` aligning in a way that requires `3 * sizeof(widest type in Device)` bytes of storage. That wasn't a problem before as both device type and index were `int8_t`.

albanD commented on c10/cuda/CUDAMacros.h:
```diff
@@ -41,11 +41,11 @@
  * The maximum number of GPUs that we recognizes. Increasing this beyond the
  * initial limit of 16 broke Caffe2 testing, hence the ifdef guards.
- * This value cannot be more than 255 because our DeviceIndex is a uint8_t.
+ * This value cannot be more than 32767 because our DeviceIndex is a int16_t.
```
Ho wow, I missed the fact that this was actually signed when I wrote this. Good update!

albanD commented on torch/csrc/jit/runtime/argument_spec.h:
```diff
@@ -58,18 +59,18 @@ struct ArgumentInfo {
   unsigned requires_grad_ : 1;
   unsigned : 5;
   unsigned dim_ : 8;
-  unsigned device_ : 8;
```
Huh, what does that mean?

tringwald replied:
I moved `device_` to the back for alignment reasons.

albanD replied:
From playing with it a bit, these padding don't make any difference since we're using bitfield which does optimal padding already.
Can you please juste remove them all?

albanD commented on torch/csrc/jit/runtime/argument_spec.h:
```diff
@@ -58,18 +59,18 @@ struct ArgumentInfo {
+  unsigned : 8;
   unsigned type_ : 8;
   unsigned dev_type_ : 16;
-  unsigned : 16;
```
What do these value represent?

tringwald replied:
I think this one is just for padding.

albanD reviewed (commented):
Sounds pretty cool!

albanD commented on torch/csrc/jit/runtime/argument_spec.h:
```diff
@@ -223,8 +224,8 @@ struct CompleteArgumentInfoPOD {
   unsigned requires_grad : 1;
-  signed device : 14;
-  unsigned dev_type : 16;
+  unsigned dev_type : 14;
```
You're reducing dev_type from 16 to 14 here!
Also we can keep the current order, since bitfield gets nicely packed anyways

tringwald replied:
I had to steal 2 bits from somewhere as the whole struct needs to fit into 64 bit. `DeviceType` is an `int8_t` anyway (with only 21 types being in use), so I don't think it matters.

tringwald replied:
In my local environment, the static assert is triggered when I swap `device` and `dev_type` back.

albanD replied:
In that case, you you make it `signed dev_type: 8;` to match the reality of things?

cyyever commented:
@albanD We need to search and replace all "std::numerical_limit<c10::DeviceIndex>::max()" with C10_COMPILE_TIME_MAX_GPUS

tringwald commented:
Maybe we should just replace all `C10_COMPILE_TIME_MAX_GPUS` with the new `C10_MAX_NUM_DEVICES` if we want to keep the device count consistent for all device types.

cyyever commented:
> Maybe we should just replace all `C10_COMPILE_TIME_MAX_GPUS` with the new `C10_MAX_NUM_DEVICES` if we want to keep the device count consistent for all device types.

It's better to use a single macro, and it is even better to change it into a constexpr variable to avoid possible upgrading issues in the future.

tringwald commented:
> > Maybe we should just replace all `C10_COMPILE_TIME_MAX_GPUS` with the new `C10_MAX_NUM_DEVICES` if we want to keep the device count consistent for all device types.
> 
> It's better to use a single macro, and it is even better to change it into a constexpr variable to avoid possible upgrading issues in the future.

Something like this maybe?

```cpp
// c10/core/Device.h
namespace c10 {
    constexpr DeviceIndex MAX_NUM_DEVICES = 512;

    struct C10_API Device final {
        // ...
    }
}
```
or
```cpp
// c10/core/Device.h
namespace c10 {
    struct C10_API Device final {
        static constexpr DeviceIndex MAX_NUM_DEVICES = 512;
    }
}
```

cyyever commented:
> ```c++
>     constexpr DeviceIndex MAX_NUM_DEVICES = 512;
> ```

IMO, the first case is better

cyyever commented on c10/core/Device.h:
```diff
@@ -209,7 +222,7 @@ struct hash<c10::Device> {
     uint32_t bits = static_cast<uint32_t>(static_cast<uint8_t>(d.type()))
             << 16 |
-        static_cast<uint32_t>(static_cast<uint8_t>(d.index()));
+        static_cast<uint32_t>(static_cast<uint16_t>(d.index()));
```
Use static_cast<uint32_t>(d.index()) directly?

tringwald replied:
This would cause problems when the index is -1 (see the code comment above).

cyyever commented:
@pytorchbot rebase

cyyever commented:
@pytorchbot label ciflow/binaries

cyyever reviewed (dismissed):
Looks good.

cyyever commented:
@tringwald Please add  validate() to DeviceIndex::set_index(DeviceIndex index)

tringwald commented:
> @tringwald Please add validate() to Device::set_index(DeviceIndex index) as we now rely on the constexpr variable for the valid range of DeviceIndex.

Sure, I should probably also add the range check to the debug asserts.

tringwald commented:
@albanD Do you want to have another look before merging? Last time this broke caffe2 testing IIRC.

tringwald commented:
@pytorchbot rebase

albanD commented on c10/core/TensorImpl.h:
```diff
@@ -3169,7 +3169,7 @@ class C10_TensorImpl_Size_Check_Dummy_Class : private TensorImpl {
   // This is a 32-bit system
   static constexpr bool check_sizes() {
-    constexpr size_t tsize = 20 * sizeof(int64_t);
+    constexpr size_t tsize = 21 * sizeof(int64_t);
```
So we have to pay a whole new int64_t here? we didn't have enough padding space for this one byte?

albanD commented:
Some small questions about struct sizes.
I'll also import to double check.

albanD reviewed (dismissed):
We are preserving MAX_NUM_DEVICES in fbcode so I don't expect it will break stuff internally. Let's merge as is!

albanD commented:
@pytorchbot merge

albanD commented:
@pytorchbot merge -i

All failures look unrelated.

albanD commented:
@pytorchbot merge -f

I'm still pretty sure they're unrelated :D

albanD commented:
@pytorchbot merge -f "All failures are unrelated but -i is not enough"

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your change but I think the Window failures are legit as they are failing now in trunk, i.e. https://hud.pytorch.org/pytorch/pytorch/commit/450339ab2d9465276e430d89010972838ba2af56' -c ignoredsignal 

Trunk is very red atm, so I'm reverting changes to restore its sanity.  If the failures are not related, I will reland the PR.

tringwald commented:
I will rebase once #120315 has landed and see if the Windows problems persist.

huydhn commented:
@pytorchbot rebase

huydhn commented:
I'm still seeing the same failures on Windows https://hud.pytorch.org/pr/119639 so I think they are related, but not sure about their links to this change yet.  I could try ssh into a Windows runner to see if I can reproduce the failures

albanD commented:
@pytorchbot rebase

rebasing now that main is green again to get clean signal

@crcrpar any context on https://github.com/pytorch/pytorch/blob/main/test/test_torch.py#L5859-L5861 and what were the failure? This PR seems to fix them but not sure why haha

crcrpar commented:
> @crcrpar any context on https://github.com/pytorch/pytorch/blob/main/test/test_torch.py#L5859-L5861 and what were the failure? This PR seems to fix them but not sure why haha

oh I almost don't remember, I guess it was numerical

cyyever commented on aten/src/ATen/native/ForeachUtils.h:
```diff
@@ -252,10 +252,16 @@ using IndicesT = std::vector<size_t>;
+  std::size_t operator()(const DeviceDtypeKey& k) const noexcept {
+    static std::hash<at::Device> device_hasher;
+    static std::hash<at::ScalarType> scalar_type_hasher;
+    return device_hasher(k.first) ^ scalar_type_hasher(k.second);
```
Would it be clearer to use 
```
std::hash(k.first) ^ std::hast(k.second)
```
?

tringwald replied:
Sure, this is just debugging code for now. Once the pipeline is green, I will clean up the code and re-add the checks I removed.

tringwald commented:
The Windows problems seem to be caused by this hash function: https://github.com/pytorch/pytorch/blob/834c7a1d3ea07878ad87d127ee28606fc140b552/aten/src/ATen/native/utils/ParamsHash.h#L13-L28

We call it on `c10::Device` for insertion into a `std::unordered_map`. My guess is, that now that `c10::Device` contains an `int8_t` and an `int16_t`, a padding byte is inserted, which VS2019 handles differently than  gcc/clang. The hash iterates over the raw bytes, so the padding is also part of the hash. This might cause keys of the `map` to randomly not match even if their values are equal.

cyyever commented:
> The Windows problems seem to be caused by this hash function:
> 
> https://github.com/pytorch/pytorch/blob/834c7a1d3ea07878ad87d127ee28606fc140b552/aten/src/ATen/native/utils/ParamsHash.h#L13-L28
> 
> We call it on `c10::Device` for insertion into a `std::unordered_map`. My guess is, that now that `c10::Device` contains an `int8_t` and an `int16_t`, a padding byte is inserted, which VS2019 handles differently than gcc/clang. The hash iterates over the raw bytes, so the padding is also part of the hash. This might cause keys of the `map` to randomly not match even if their values are equal.

Is it possible to place DeviceIndex before DeviceType in Device to avoid it?

tringwald commented:
It will be padded either way.

cyyever commented:
> It will be padded either way. We could `#pragma pack` it, but that has some penalties we probably don't want for such a core structure.

A quick solution is to specify ParamsHash for Device. I think the template has serious design flaws.

tringwald commented:
We probably shouldn't use `ParamsHash` at all, due to the padding behavior. I have rewritten the `unordered_map` hashing logic to use the already existing hash function in `Device.h`.

Apparently, this also isn't the first time someone encountered this problem ... https://github.com/pytorch/pytorch/blob/5d71ba688563ef491bb28d47c493ec6fc7791da2/aten/src/ATen/native/quantized/cudnn/Conv.cpp#L92-L96

tringwald committed:
Switched hashing logic for c10::Device in multiple places, added a warning for unexpected behavior.

albanD reviewed (dismissed):
Ho wow!
Thanks for investigating this, it is a weird one!

tringwald commented:
@pytorchbot merge

kit1980 commented:
I'm going to revert this.
Lots of Meta-internal build failures in the code that uses PyTorch (not PyTorch itself):
`error: non-constant-expression cannot be narrowed from type 'c10::DeviceIndex' (aka 'short') to 'int8_t' (aka 'signed char') in initializer list [-Wc++11-narrowing]`

To reland this, someone from Meta needs to help.

kit1980 commented:
@pytorchbot revert -m "breaking internal builds, see D54286923" -c ghfirst

tringwald commented:
Unfortunately, this is now blocking some XPU work that was based on the new `DeviceIndex` and `MAX_NUM_DEVICES` constant. Would it work out for Meta internal builds if we hide the new `int16_t` `DeviceIndex` behind a preprocessor flag and the internal builds use the old `int8_t` index?

tomasikp commented:
Tracking this PR

larrysingh commented:
Tracking this issue

tringwald commented:
Can someone from Meta comment on how to proceed here? Last time, when a similar change broke Caffe2, we hid it behind `#ifdef FBCODE_CAFFE2`. Would this be applicable here too?

kit1980 commented:
@tringwald I'll investigate and let you know.

kit1980 commented:
@pytorchbot rebase

kit1980 commented:
@tringwald I think I figured out what to change internally.
Can you resolve the conflicts/rebase this PR?

tringwald commented:
> @tringwald I think I figured out what to change internally. Can you resolve the conflicts/rebase this PR?

Great, thanks for investigating! I've rebased onto main and resolved the conflict.

tringwald commented:
@pytorchbot merge

tringwald commented:
@kit1980 is it fine to merge this PR now or do we need to coordinate the merge with Meta-internal changes?

kit1980 commented:
@tringwald I'm going to merge myself simultaneously internally and on Github.
Hopefully today.

kit1980 commented on caffe2/core/context_gpu.cu:
```diff
@@ -178,8 +178,8 @@ static std::unordered_map<void*, uint8_t> g_cuda_device_affiliation;
 static std::unordered_map<void*, long> g_size_map;
-static std::vector<long> g_total_by_gpu_map(C10_COMPILE_TIME_MAX_GPUS, 0);
-static std::vector<long> g_max_by_gpu_map(C10_COMPILE_TIME_MAX_GPUS, 0);
+static std::vector<long> g_total_by_gpu_map(c10::Device::MAX_NUM_DEVICES, 0);
```
In line 176, shouldn't `static std::unordered_map<void*, uint8_t> g_cuda_device_affiliation;` also be updated to uint16_t?

tringwald replied:
Good catch, I guess we should use `DeviceIndex` here too.

kit1980 commented:
@tringwald 
Because `DeviceIndex` used to be `int8_t` before this PR, int8 was supported primitive type here: https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h#L75 

So now `int8_t` needs to be added to `supported_primitive_arg_types`, otherwise it's a backward-compatibility breaking change.

Similarly, in https://github.com/pytorch/pytorch/blob/0d6cac3d1aa5c0569010064f08e8977227483a84/aten/src/ATen/core/op_registration/infer_schema.h#L40 int16 needs to be added without removing int8.

Does this make sense? (a small internal change will still be needed after updating this, so I'll habe to merge).
cc @albanD

tringwald committed:
Preserved backwards compatibility for int8_t primitives, changed CUDA device affiliation map from uint8_t to DeviceIndex.

albanD reviewed (approved):
Sounds good!

kit1980 commented:
There is still one internal failure I'm trying to figure out.

```
terminate called after throwing an instance of 'c10::Error'
  what():  Type signed char could not be converted to any of the known types.
```

kit1980 commented:
Also 
```
ERROR: test_int16_device_index_cuda (test_utils.TestDeviceUtilsCUDA)
RuntimeError: Device index must be between 0 and 15 inclusively.
```

kit1980 commented:
Also
```
ERROR: test_int16_device_index_cuda (test_utils.TestDeviceUtilsCUDA)
RuntimeError: Device index must be between 0 and 15 inclusively.
```

tringwald commented:
> Also
> 
> ```
> ERROR: test_int16_device_index_cpu (test_utils.TestDeviceUtilsCPU)
> RuntimeError: Device index must be between 0 and 15 inclusively.
> ```
> 
> The actual value of index was 500 for some reason...

The problem here is that we hardcode `MAX_NUM_DEVICES` to be `16` when `FBCODE_CAFFE2` is set. This was the result of some previous changes w.r.t. the device index, maybe @albanD knows some more details on that. 
I will add some changes to disable the test for Meta-internal test runs.

tringwald committed:
Removed ifdef guards for Meta-internal DeviceIndex. This should not be a problem anymore when the internal code base also uses int16_t.

albanD commented:
We can try that. If not, there is a flag in test common_utils to be able to skip the test if it is in fbcode.

kit1980 commented:
@pytorchbot rebase

kit1980 commented:
@pytorchbot rebase

tringwald committed:
Fixed formatting after rebase.

tringwald commented:
@kit1980 I rebased on main and fixed the merge conflict

tringwald commented:
@pytorchbot rebase

tringwald commented:
All CI failures seem unrelated. Unfortunately, neither main nor viable/strict are stable right now.

kit1980 commented:
I'll try the latest variant on Friday.

kit1980 commented:
@pytorchbot rebase

kit1980 commented:
@pytorchbot rebase

tringwald commented:
I rebased on `main` to fix the merge conflicts caused by the now removed `caffe2/contrib` package. Do we need to wait for the full removal of `caffe2` (#122527) to get this merged? @kit1980

cyyever commented:
@tringwald It may take a long time to fully remove caffe2..

tringwald committed:
Cleanup after rebase. Fixed formatting.

jataylo commented:
Hey @tringwald @cyyever @albanD @kit1980 @huydhn 

Is there a plan for when this change can be merged? The 120 GPU limit is restrictive.

albanD commented on caffe2/contrib/nccl/cuda_nccl_op_gpu.cc:
I guess you didn't mean to create this new file?

albanD reviewed (commented):
Ho that one dropped yes.
Only a small nit but sounds good otherwise I think

cyyever committed:
Delete caffe2/contrib/nccl/cuda_nccl_op_gpu.cc

It was removed before.

albanD reviewed (changes requested):
A few of the APIs being removed are used, so this needs manual fixing before it can land.

tringwald commented:
> A few of the APIs being removed are used, so this needs manual fixing before it can land.

Are you referring to Meta-internal code? I think @kit1980 has done some work on the Meta side.

kit1980 commented:
> I think @kit1980 has done some work on the Meta side.

I could not figure out one item and then caffe2 desync caused additional issues.
